### PR TITLE
Added support for this referencing inside firebase funcs

### LINKED ISSUE #399: Wrong type of 'this' inside firestore() context
I am getting typing errors when trying to access props/methods of Vue instance inside of `firestore` function. In the example below, `uid` is a prop and I get no typing errors when accessing from `created` lifecycle hook, but it is not defined inside firestore function.

```ts
export default Vue.extend({

  props: ['uid'],

  created() {
    console.log({ uid: this.uid }) // this.uid is fine
  },

  firestore() {
    // Error: "Property 'uid' doesn't exist on type Vue"
    const uid = this.uid 
    return {
      user: db.doc(`users/${uid}`),
    }
  },
})
```

This can resolve the issue, but shouldn't be necessary:
```ts
const uid = (this as any).uid 
```

### PR DESCRIPTION
**Pull request by trickstival**
## The problem
When using typescript, it is not possible to access properties defined in `this` context. It warns that the defined props are not available in the scope with the message `Error: "Property 'nameOfProp' doesn't exist on type Vue"`.

## Proposed solution
I augmented `VueConstructor` type so it passes default typings around on `extend` function (which doesn't happen in Vue itself) and passed a CombinedInstance type as `this` for the `firestore` and `firebase` functions.

I'm not sure if that's a good approach, cause it messes up with Vue module definition.
If you know a better way of implementing this, pls share.

## Still not fixed
- I haven't figured out yet how to infer methods, data and computed properties types. For now, it only infers prop types and treats everything else as `any`

fix #399

### CODE REVIEW

**Comment by TomKaltz** [comment]
What's the status here?

**Comment by trickstival** [comment]
I think there's a coverage problem, but I'm still waiting for the review

**Comment by posva** [comment]
I'm not sure if there is a way to also type the methods and other properties. Maybe using the type of a method in a Vue instance we can infer `this` correctly. Having a partial type only is not enough

**Comment by trickstival** [comment]
I'm not sure if I get what you mean by using the type of a method in a Vue instance. The typing code is getting pretty hard to follow.

**Comment by posva** [comment]
> The typing code is getting pretty hard to follow.

yeah, I know...

I mean that doing `this.myMethod()` should also work but currently it doesn't

**Comment by posva** [comment]
Thank you for this and sorry I never came back again
This ended up changing quite a lot and this PR isn't relevant anymore. The progress of the latest release can be followed at https://github.com/vuejs/vuefire/issues/1241
